include/nbl/builtin/hlsl/impl/tgmath_impl.hlsl

```diff
@@ -0,0 +1,328 @@
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct exp_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct exp2_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct log_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct abs_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct cos_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct sin_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct acos_helper;
+template<typename T NBL_STRUCT_CONSTRAINABLE>
+struct sqrt_helper;
+template<typename T, typename U NBL_STRUCT_CONSTRAINABLE>
+struct lerp_helper;
+
+#ifdef __HLSL_VERSION
+
+#define AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(HELPER_NAME, SPIRV_FUNCTION_NAME, RETURN_TYPE)\
+template<typename T> NBL_PARTIAL_REQ_TOP(always_true<decltype(spirv::SPIRV_FUNCTION_NAME<T>(experimental::declval<T>()))>)\
+struct HELPER_NAME<T NBL_PARTIAL_REQ_BOT(always_true<decltype(spirv::SPIRV_FUNCTION_NAME<T>(experimental::declval<T>()))>) >\
+{\
+	using return_t = RETURN_TYPE;\
+	static inline return_t __call(const T arg)\
+	{\
+		return spirv::SPIRV_FUNCTION_NAME<T>(arg);\
+	}\
+};
+
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(cos_helper, cos, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(sin_helper, sin, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(abs_helper, sAbs, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(abs_helper, fAbs, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(sqrt_helper, sqrt, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(log_helper, log, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(exp2_helper, exp2, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(exp_helper, exp, T)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(floor_helper, floor, T)
+#define ISINF_AND_ISNAN_RETURN_TYPE conditional_t<is_vector_v<T>, vector<bool, vector_traits<T>::Dimension>, bool>
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(isinf_helper, isInf, ISINF_AND_ISNAN_RETURN_TYPE)
+AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER(isnan_helper, isNan, ISINF_AND_ISNAN_RETURN_TYPE)
+#undef ISINF_AND_ISNAN_RETURN_TYPE 
+#undef AUTO_SPECIALIZE_TRIVIAL_CASE_HELPER
+
+template<typename T, typename U> NBL_PARTIAL_REQ_TOP(always_true<decltype(spirv::pow<T>(experimental::declval<T>(), experimental::declval<T>()))>)
+struct pow_helper<T, U NBL_PARTIAL_REQ_BOT(always_true<decltype(spirv::pow<T>(experimental::declval<T>(), experimental::declval<T>()))>) >
+{
+	using return_t = T;
+	static inline return_t __call(const T x, const T y)
+	{
+		return spirv::pow<T>(x, y, a);
+	}
+};
+
+template<typename T, typename U> NBL_PARTIAL_REQ_TOP(always_true<decltype(spirv::fMix<T>(experimental::declval<T>(), experimental::declval<T>(), experimental::declval<U>()))>)
+struct lerp_helper<T, U NBL_PARTIAL_REQ_BOT(always_true<decltype(spirv::fMix<T>(experimental::declval<T>(), experimental::declval<T>(), experimental::declval<U>()))>) >
+{
+	using return_t = conditional_t<is_vector_v<T>, vector<bool, vector_traits<T>::Dimension>, bool>;
+	static inline return_t __call(const T x, const T y, const U a)
+	{
+		return spirv::fMix<T>(x, y, a);
```

Comment:
call with all template params!

Reply:
from SPIR-V spec:
![image](https://github.com/user-attachments/assets/d42bcfad-b417-4c76-8b71-41d0505020f8)
the template parameter `U` is for C++ side specializations

Reply:
if you can't call `fMix` with `U a` then the signature of `__call` here should have `const T a`

Reply:
i don't think i get it, you want to restrict the helper to only one template parameter? i was requested to make `mix` compatible with bool as the 3rd argument

Reply:
note that this function has been moved to the `intrinsics_impl.hlsl` file